### PR TITLE
TMA-747: optimize amount of requests for executing MUFs

### PR DESCRIPTION
**Pull request by kubamahnert**

### CODE REVIEW

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 1]
```diff
@@ -160,7 +160,14 @@ def call(params)
 
             working_client_ids = []
 
-            filters.group_by { |u| u[multiple_projects_column] }.flat_map do |client_id, new_filters|
+            grouped_filters = filters.group_by { |u| u[multiple_projects_column] }
+            project_users = {}
+            grouped_filters.flat_map do |client_id, _|
+              project = domain.clients(client_id, data_product).project
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'domain.clients(client_id, data_product)' 3 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'domain.clients(client_id, data_product).project' 2 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 2]
```diff
@@ -160,7 +160,14 @@ def call(params)
 
             working_client_ids = []
 
-            filters.group_by { |u| u[multiple_projects_column] }.flat_map do |client_id, new_filters|
+            grouped_filters = filters.group_by { |u| u[multiple_projects_column] }
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'filters.group_by { |u| u[multiple_projects_column] }' 3 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'filters.group_by' 3 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'u[multiple_projects_column]' 3 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Has the variable name 'u' (UncommunicativeVariableName)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 3]
```diff
@@ -160,7 +160,14 @@ def call(params)
...
+            project_users = {}
+            grouped_filters.flat_map do |client_id, _|
+              project = domain.clients(client_id, data_product).project
+              project_users[project.uri] = project.users unless project_users[project.uri]
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'project.uri' 3 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'project_users[project.uri]' 2 times (DuplicateMethodCall)

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/624/ : SUCCESS in 10s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/418/ : SUCCESS in 1m 57s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/726/ : SUCCESS in 1m 55s

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 4]
```diff
@@ -160,7 +160,14 @@ def call(params)
 
             working_client_ids = []
 
-            filters.group_by { |u| u[multiple_projects_column] }.flat_map do |client_id, new_filters|
+            grouped_filters = filters.group_by { |u| u[multiple_projects_column] }
+            project_users = {}
+            grouped_filters.flat_map do |client_id, _|
```
Comment (panjan) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Why iterate over grouped_filters twice? Why flat_MAP if the result is not used?

File: lib/gooddata/models/user_filters/user_filter_builder.rb [hunk 5]
```diff
@@ -365,11 +365,11 @@ def self.execute_mufs(user_filters, options = {})
       client = options[:client]
       project = options[:project]
       ignore_missing_values = options[:ignore_missing_values]
-      users_must_exist = options[:users_must_exist] == false ? false : true
+      users_must_exist = options[:users_must_exist]
       dry_run = options[:dry_run]
       project_log_formatter = GoodData::ProjectLogFormatter.new(project)
 
-      project_users = project.users
+      project_users = options[:project_users] || project.users
```
Comment (panjan) [inline lib/gooddata/models/user_filters/user_filter_builder.rb]: Can you write a test for this please?

File: lib/gooddata/models/project_log_formatter.rb [hunk 6]
```diff
@@ -165,11 +164,9 @@ def log_user_filter_results(results, user_filters)
         if status == :successful
           filter_uris = user_filters[user_profile_url].map(&:uri)
           if operator == :create && GoodData.logger.info?
-            readable_user_login = users_cache[user_profile_url] || user_profile_url
-            GoodData.logger.info "Created user-filter=#{filter_uris} for user=#{readable_user_login} in project=#{@project.pid}"
+            GoodData.logger.info "Created user-filter=#{filter_uris} for user=#{user_profile_url} in project=#{@project.pid}"
```
Comment (yenkins) [inline lib/gooddata/models/project_log_formatter.rb]: Calls '@project.pid' 4 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/models/project_log_formatter.rb]: Calls 'gooddata.logger' 8 times (DuplicateMethodCall)

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/633/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/426/ : SUCCESS in 34s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/734/ : SUCCESS in 1m 49s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/637/ : SUCCESS in 20s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/429/ : SUCCESS in 49s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/737/ : SUCCESS in 2m 09s

**Comment by panjan** [comment]
ok to merge

File: lib/gooddata/models/project.rb [hunk 7]
```diff
@@ -1553,26 +1553,27 @@ def uri
...
+      offset = opts[:offset] || 0
+      limit = opts[:limit] || 1_000
+
+      loop do
+        tmp = client.get("/gdc/projects/#{pid}/users", params: { offset: offset, limit: limit })
+        tmp['users'].each do |user_data|
+          user = client.create(GoodData::Membership, user_data, project: self)
+
+          if opts[:all]
+            res << user
```
Comment (yenkins) [inline lib/gooddata/models/project.rb]: Calls 'res << user' 3 times (DuplicateMethodCall)

File: lib/gooddata/models/project.rb [hunk 8]
```diff
@@ -1553,26 +1553,27 @@ def uri
...
-            else
-              y << user if user && user.enabled?
-            end
+      res = []
+      offset = opts[:offset] || 0
+      limit = opts[:limit] || 1_000
+
+      loop do
+        tmp = client.get("/gdc/projects/#{pid}/users", params: { offset: offset, limit: limit })
+        tmp['users'].each do |user_data|
```
Comment (yenkins) [inline lib/gooddata/models/project.rb]: Calls 'tmp['users']' 2 times (DuplicateMethodCall)

File: lib/gooddata/models/user_filters/user_filter_builder.rb [hunk 9]
```diff
@@ -140,10 +140,12 @@ def self.verify_existing_users(filters, options = {})
 
     def self.create_label_cache(result, options = {})
       project = options[:project]
+      project_labels = project.labels
+
 
       result.reduce({}) do |a, e|
         e[:filters].map do |filter|
-          a[filter[:label]] = project.labels(filter[:label]) unless a.key?(filter[:label])
+          a[filter[:label]] = (project_labels.find { |l| l.uri == filter[:label] } || project.labels(filter[:label])) unless a.key?(filter[:label])
```
Comment (yenkins) [inline lib/gooddata/models/user_filters/user_filter_builder.rb]: Calls 'filter[:label]' 4 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/models/user_filters/user_filter_builder.rb]: Contains iterators nested 3 deep (NestedIterators)

Comment (yenkins) [inline lib/gooddata/models/user_filters/user_filter_builder.rb]: Has the variable name 'l' (UncommunicativeVariableName)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/647/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/435/ : SUCCESS in 2m 07s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/743/ : FAILURE in 1m 54s

File: lib/gooddata/models/user_filters/user_filter_builder.rb (hunk 9 above)
Comment (panjan) [inline lib/gooddata/models/user_filters/user_filter_builder.rb]: Would it be better to eliminate the `||` branch by this? `(project_labels.find { |l| l.uri == filter[:label] || l.id == filter[:label] }`

**Comment by gdgate** [comment]
Merge Failed.

This change or one of its cross-repo dependencies was unable to be automatically merged with the current state of its repository. Please rebase the change and upload a new patchset.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/662/ : SUCCESS in 10s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/462/ : SUCCESS in 2m 09s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/770/ : SUCCESS in 2m 04s

**Comment by gdgate** [comment]
Merge Failed.

This change or one of its cross-repo dependencies was unable to be automatically merged with the current state of its repository. Please rebase the change and upload a new patchset.

**Pull request closed without merging**